## Classifcation confidence check

We decided to check what is the average level of confidence for prediction when predictions are faulty as this values could be used to filter out false positives and reduce the cost of errors. Script called `confidencesim.py` was used in order to perform 50 full training and classification runs. 

In [22]:
import pandas as pd
import numpy as np

In [18]:
df = pd.read_csv('../data/confidence_sim.csv')

In [19]:
df.head()

0_confidence  1_confidence  predicted_class  real_class
0      0.686944      0.313056                0           1
1      0.269016      0.730984                1           0
2      0.926314      0.073686                0           1
3      0.388214      0.611786                1           0
4      0.520184      0.479816                0           1

In [23]:
df['error_type'] = np.where(df['predicted_class']==1, 'FP', 'FN')

In [24]:
df.head()

0_confidence  1_confidence  predicted_class  real_class error_type
0      0.686944      0.313056                0           1         FN
1      0.269016      0.730984                1           0         FP
2      0.926314      0.073686                0           1         FN
3      0.388214      0.611786                1           0         FP
4      0.520184      0.479816                0           1         FN

In [28]:
df.drop(['predicted_class', 'real_class'], axis=1, inplace=True)
df.groupby('error_type').mean()

0_confidence  1_confidence
error_type                            
FN              0.704757      0.295243
FP              0.350733      0.649267

On average:
- when FN prediction is made, the confidence does not exceed 71%,
- when FP prediction is made, the confidence does not exceed 65%.

Based on that we can assume that positive classification with confidence lower than 0.7 are likely to be false positive. With the false positive error cost being 5 times higher than the false negative error cost it seems reasonable to swap all these predictions to 0.